ci: automate major release tasks

goosewobbler opened the pull request:
Closes #959.

This includes:

- [x] Automated update of dependabot.yml
- [x] Reusable workflows
- [x] Updated development documentation to reflect new release process
- [x] Automated branch lifecycle management
	- [x] Automatic creation of branches during major releases
		- [x] feature
		- [x] maintenance
	- [x] Automatic archiving (deletion) of old maintenance branches
- [x] Enhanced automation for major releases
	- [x] 	Automatic dependabot.yml updates for new maintenance branch
	- [x] 	Automatic merge of feature branch to main
- [x] Updated release workflows to use dynamic branch resolution
	- [x] feature → automatically resolves to next major version branch (e.g., `feature/v9`)
	- [x] main → current stable version
	- [x] maintenance → automatically resolves to previous version branch (e.g., `v7`~~`.x`~~)
- [x] Enhanced backporting process
	- [x] Dynamic version detection from package.json
	- [x] Interactive version confirmation when auto-detection fails
	- [x] Improved error handling and user feedback
	- [x] Clear version information display during execution
- [x] Rework of labels and strategy for usage 
- [x] Updated documentation for all of the above

goosewobbler commented:
WIP as I am evaluating the different approaches for updating `dependabot.yml`

mato533 commented:
@goosewobbler 
My idea is using `yp` to rewrite yaml config as simple way. I hope you like it.

https://github.com/marketplace/actions/yq-portable-yaml-processor

```
- name: Replace dependabot config
  uses: mikefarah/yq@v4.45.1
  with:
    cmd: yq -i '.updates[1].target-branch = "v8"' '.github/dependabot.yml'
```

mato533 commented:
@goosewobbler 
I have created the PR to this PR. I hope it could be your help.

goosewobbler commented:
@mato533 sorry I didn't get time to look at this recently.  The yq approach for updating the dependabot config seems sound, and I like the workflow reuse.  I'll close this PR and use yours as a base for this piece of work.

mato533 commented:
Good! I'm glad you like it.

goosewobbler commented:
Targetting `8.0.1` for this

mato533 commented on .github/workflows/release-shared.yml:
```diff
@@ -0,0 +1,356 @@
+          tag_name: ${{ steps.push-tags.outputs.RELEASE_TAG }}
+          token: $GITHUB_TOKEN
+
+      # Update branch protection rules for new branches during major releases
+      - name: 🔒 Update branch protection rules
+        if: ${{ contains(inputs.releaseVersion, 'major') && !inputs.dryRun }}
+        env:
+          GITHUB_TOKEN: ${{ secrets.github-token }}
+          NEXT_LTS_BRANCH: ${{ needs.calculate.outputs.next-lts-branch }}
+          REPO_OWNER: webdriverio-community
+          REPO_NAME: wdio-electron-service
+        run: pnpx tsx scripts/update-branch-protection.ts
```
This step has the problem...

To update branch protection rules, the following permission is required.
- "Administration" repository permissions (write) [see this docs](https://docs.github.com/en/rest/branches/branch-protection?apiVersion=2022-11-28#update-branch-protection--fine-grained-access-tokens)

But the GitHub Actions could not be set the permission of one.
[See this docs](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#defining-access-for-the-github_token-scopes-1)

```suggestion
```

### Alternatives
We can create branch protection rules using pattern (`v[0-9]*`) like this. 
![image](https://github.com/user-attachments/assets/9ef98d9c-a3e9-4cc4-a414-48a3a1125db5)

But due to `fnmatch` limitations, it is protect all branch of named start with  v + numeric (e.g v9, v10, v10xx)
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/creating-rulesets-for-a-repository#using-fnmatch-syntax

goosewobbler replied:
You're right about the perms here, I didn't test this (rapid iteration on this and I wanted early feedback) but in any case I think your suggestion regarding the target branches is more elegant, so we can remove this

mato533 commented on scripts/update-branch-protection.ts:
If intend to use only on the GitHub Actions, we could remove this script.

mato533 reviewed (changes requested):
@goosewobbler 
I posted a comment about the points I've noticed during testing in my forked repo.
It would be nice if you could check.

mato533 commented:
@goosewobbler 
Let me confirm something about branch naming conventions.

Let's assume that the assumption is that the current version is 8.0.0.
Is the naming of each of the following branches correct, I am wondering if the feature branch should include the version? (as in `feature/v9`)
- Name of the feature branch : `feature`
- Name of the main branch : `main` (there are no other options)
- Name of the maintenance branch : `v7`

mato533 commented:
@goosewobbler 

Additionally, there are people who waiting #957 were released, so we should better to release 8.0.1. 

And improvements in CI are not directly relevant to end-users.

goosewobbler commented:
@mato533 you're right, I released 8.0.1 - I thought this PR would be a quick one to get in but there are a lot of changes here.

mato533 commented:
@goosewobbler
Thank you! It's nice!

And I have create the PR toward this PR. (#1004)
This implements following steps.

> - Automated branch lifecycle management
>      - Automatic creation of maintenance branch during major releases

goosewobbler commented on .github/workflows/release-shared.yml:
```diff
@@ -0,0 +1,356 @@
+          if [[ "${TARGET_BRANCH}" == "feature/"* ]]; then
+            # For feature branches, use as-is
+            echo "target-branch=${TARGET_BRANCH}" >> $GITHUB_OUTPUT
+          elif [[ "${TARGET_BRANCH}" == "v"*".x" ]]; then
```
Ah yes, this is an artifact from the previous approach of having the maintenance branch as `v*`

goosewobbler commented on .github/workflows/release-shared.yml:
```diff
@@ -0,0 +1,356 @@
+            if [[ "$pkg_json" != *"node_modules"* ]]; then
+              PACKAGE_JSONS+=("$pkg_json")
+            fi
+          done < <(find packages -name "package.json" -not -path "*/node_modules/*")
```
Good catch, I forgot about the extra package.jsons we have in there for the subpath import fixes

goosewobbler commented:
@mato533 I addressed your feedback, what do you think about the labelling strategy?  

We have new labels defining release track, version timing, version increment and priority - the usage of these is documented in the [release management doc](https://github.com/webdriverio-community/wdio-electron-service/blob/sm/automate-major-release/docs/release-management.md)

mato533 reviewed (changes requested):
I have putted some change requests. please check them.

goosewobbler commented on .github/workflows/release-shared.yml:
```diff
@@ -0,0 +1,364 @@
+      - name: Check Release Parameters
+        env:
+          IS_MAJOR: ${{ contains(inputs.releaseVersion, 'major') }}
+          IS_FEATURE: ${{ contains(inputs.branch, 'feature') }}
+          IS_MAINTENANCE: ${{ contains(inputs.branch, '.x') }}
```
more artifacts of the previous approach, gotta love AI.  We're not being replaced any time soon.

mato533 commented:
@goosewobbler 
When I saw the commit of [2a4e74d](https://github.com/webdriverio-community/wdio-electron-service/pull/960/commits/2a4e74db0a1cedcb4c0353b051b3a71ea902518f), It may  better to restrict to run entire workflow only community repo.

Specifically, add the repo name check to `check` job.

(for now, I'm running my forked repo for testing, but basically it is better to avoid run on not community repo.)

mato533 commented on .github/workflows/release-shared.yml:
```diff
@@ -0,0 +1,439 @@
+          prerelease: ${{ contains(inputs.releaseVersion, 'pre') }}
+          repository: webdriverio-community/wdio-electron-service
+          tag_name: ${{ steps.push-tags.outputs.RELEASE_TAG }}
+          token: $GITHUB_TOKEN
```
More appropriately, the Token itself should be passed as a string, which is easier to understand, so a change is suggested.
```suggestion
          token: ${{ secrets.github-token }}
```

goosewobbler replied:
I think this has been resolved in the split

mato533 commented:
The size of the Release job is also considerably larger and less prospective, so it might be a good idea to separate the responsibilities of the Release job itself and the jobs related to Major Version UP.

mato533 commented on .github/workflows/release-shared.yml:
```diff
@@ -0,0 +1,364 @@
+        # Only run in the official repository and not during dry runs
+        # This prevents milestone creation in forks (which would fail due to permissions)
+        # and ensures no milestones are created during dry runs (which should be read-only)
+        if: ${{ github.repository == 'webdriverio-community/wdio-electron-service' && !inputs.dryRun }}
```
@goosewobbler 
This suggestion should be considered with [this comment](https://github.com/webdriverio-community/wdio-electron-service/pull/960#issuecomment-2701266120)

goosewobbler replied:
Updated

goosewobbler commented:
@mato533 I reworked the workflow into modules and stopped both release workflows from being used outside the main repo...

goosewobbler commented:
Getting some flaky E2Es on Ubuntu, I remember these particular tests were problematic a long time ago.  None of these CI changes should have any bearing on the tests.

mato533 commented:
@goosewobbler 
> Getting some flaky E2Es on Ubuntu only, I remember these particular tests were problematic a long time ago. None of these CI changes should have any bearing on the tests. Could also be some kind of github runner / image glitch

yes, I recognised it, and I have created the issue #1005 
I have same understanding that there are no relations between this PR and the problem.
It is required the investigation.

mato533 commented:
@goosewobbler 
It's great! but we are stuck with GitHub Actions restrictions.....

It is not supported the subdirectories about reusable workflow.....(Many times I have thought it would be wonderful if I could do that.)

[See Docks](https://docs.github.com/en/actions/sharing-automations/reusing-workflows#creating-a-reusable-workflow)
> Subdirectories of the workflows directory are not supported.

- Not supported
    .github/workflows/release-modules/build-verify.yml
    
- Suppoted
    .github/workflows/build-verify.yml

#### Alternatives
A prefix can be added so that the main workflow and the associated Reusablle Workflow can be recognised.

e.g.
release.yml (main workflow)
release-verify.yml (called from main workflow. means reusable workflow)
release-calculate.yml (called from main workflow. means reusable workflow)
release-post.yml (called from main workflow. means reusable workflow)

mato533 commented:
Additionally, `setup.yml` could not be created as reusable workflow....
This is because the virtual OS (container) changes in the unit in which the workflow is executed, so, for example, the git config user.email value set in `setup.yml` will not be set when the release-tasks.yml is executed. (and therefore you will probably get an error during PUSH)

We consider it appropriate to create Steps such as those present in `setup.yml` as Actions and define them as Steps in a Job when called from a workflow.

mato533 commented:
@goosewobbler 
I have created the PR to adjust the GitHub Actions limitations.

mato533 commented:
Following conditions is confricted... 
https://github.com/webdriverio-community/wdio-electron-service/blob/8cfad8d527b38abc926f20e3cf93eeca8ef3683d/.github/workflows/%40release-shared-calculate.yml#L39-L43

https://github.com/webdriverio-community/wdio-electron-service/blob/8cfad8d527b38abc926f20e3cf93eeca8ef3683d/.github/workflows/%40release-shared-post.yml#L145-L146

mato533 commented:
The `@` could not be used as filename.... we have to remove this character...

![image](https://github.com/user-attachments/assets/6ea4ede9-75d7-4427-9ea3-c2f662813ae6)

mato533 commented:
It seems natural that the official release tag should be in the Main branch.

I thought it might be more natural to pre-process the release, merge it, commit it in the main branch and give it a tag.

<img width="715" alt="image" src="https://github.com/user-attachments/assets/173dfe9e-2c0c-44a1-a483-9d5982bd39f6" />

mato533 commented:
@goosewobbler 
Thank you for update and merging my PR!
I am testing my repo... and looking at how TAGging can be done properly. We are close to a solution and will put it together in a PR when it is ready!

goosewobbler commented:
@mato533 I moved things around and added a split between prepare and publish.  Definitely close...

mato533 commented:
@goosewobbler 
year, we are close...
turbo-version is error. I think turbo-version should be before publish....
![image](https://github.com/user-attachments/assets/526afaaa-ad9a-45fb-9155-1253515ef303)

goosewobbler commented:
A lot more went into this one than I thought initially, but it's important to have a proper automated release process.

mato533 reviewed (approved):
The normal system appears to be working properly!

mato533 commented:
@goosewobbler 
There are likely to be some precautions that you were not aware of.

I am aware that I need to manually merge the feature branch into the main branch before running this Workflow. (It is expected, and the logs will show this.) When doing this merge, the package.json will always be conflicted.
How to resolve this is that you must always specify the current version of the main branch. (calculations don't work properly).
(e.g. if main=9.x.y, feature/v10= 10.x.y.next-z, we need to make sure that the merge is done at 9.x.y)

![image](https://github.com/user-attachments/assets/59fb8c56-c001-4704-9c67-cad3ed2cba80)
